### PR TITLE
WIP: LSP goto definition for proc

### PR DESCRIPTION
**Pull request by patrislav1**
Initial PR that supports `textDocument/definition` for procs as agreed in https://github.com/nmoroze/tclint/pull/101.

### CODE REVIEW

File: src/tclint/syntax_tree.py
Comment (nmoroze) [inline src/tclint/syntax_tree.py:1]: I think I'd prefer to leave the `prev`/`next` API changes out of this PR for now. I know of at least one place in the code where `children` is updated directly which would break this, and I have another idea for how to handle `SymbolTable.get_symbol_text()`, which I think is the only consumer right now.

File: src/tclint/cli/tclsp.py [hunk 1]
```diff
@@ -330,6 +331,57 @@ def format_range(ls: TclspServer, params: lsp.DocumentRangeFormattingParams):
     ]
 
 
+@server.feature(lsp.TEXT_DOCUMENT_DEFINITION)
+def goto_definition(ls: TclspServer, params: lsp.DefinitionParams):
```
Comment (nmoroze) [inline src/tclint/cli/tclsp.py:335]: I thought of some edge cases for this feature that I might want addressed before it comes user-visible. I'm happy to do a slimmed-down PR with just the symbol table before solving these, in which case let's remove this for now (or perhaps feature flag it somehow?)

File: src/tclint/symbol_table.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    def visit_command(self, command: Command) -> None:
+        if command.routine.contents == "proc":
+            self.table.add_proc_definition(command)
+
+
+def get_symbol_text(node: Node) -> str | None:
+    """Get symbol text to look for from the node under the cursor"""
+    # TODO: where should this code live? it's not really part of a symbol table, maybe it belongs closer to the LSP stuff
+    # "proc foo" with cursor over "proc"
```
Comment (nmoroze) [inline src/tclint/symbol_table.py:51]: I think I have another plan that will simplify this (and eliminate the need for the Node API prev/next tweak).

I started thinking about this in the last PR, but I'd like to address this by adding a node type called `CommandRef`. This would wrap the first argument to a `Command` node (rather than a `BareWord`), and then your logic knows any instance of a `CommandRef` is a reference to a `proc` without having to check its predecessors.

I think encoding this information in the syntax tree might be convenient for other things in the future. 

If you're interested in trying this syntax tree addition in a follow-up PR let me know, otherwise I'm happy to make some time to do it myself.

File: src/tclint/symbol_table.py
Comment (nmoroze) [inline src/tclint/symbol_table.py:1]: Before we merge this without the full go-to definition implementation, I'd suggest adding some unit tests of this module (maybe in `tests/test_symbol_table.py`).

**Review by nmoroze — COMMENTED** [review]
Thanks for opening a slimmed-down PR! 

I added a few inline comments. Once those are addressed I think we'll have something I'm ready to merge now. Summarizing the main changes:
- Removing the first commit (the one adding `prev`/`next` to Nodes)
- Removing the addition of `goto_definition` from the LSP code for now
- Removing `get_symbol_text` from `SymbolTable`
- A few direct tests of `symbol_table.py` in place of the e2e LSP tests

After that's merged, let's do a follow-up PR adding a `CommandRef` node. I've also thought a bit about some edge cases for this feature, and we should figure out the intended behavior for these. 

## Edge cases

**Out-of-order usage**

Command definitions do not exist until `proc` is executed. I think the definition of `foo` should not resolve to the `proc` below it in this case. 

```tcl
foo ;# undefined
proc foo {} {}
```

**Conditional definition**

The definition of a `proc` may be conditional. This applies to e.g. if statements, a particularly weird example is the case of nested procs. This would be hard to resolve statically, but in this case I see some justification for having the definition to resolve (since it **may** be a valid definition):

```tcl
proc foo {} {
  proc bar {} {}
}
bar ;# undefined
foo
bar ;# now it works!
```

**Definition invalidation**

Some commands can un-define procs. It would be impossible to handle this perfectly with static analysis, but it might make sense to take this into account in some way. 

```tcl
proc foo {} {}
rename foo newfoo
foo #; undefined
```

**Invalidation via redefinition**

If another file is sourced (or a package required, or procs imported from a namespace), they may overwrite the definition of a `proc` and invalidate it.

```tcl
proc foo {} {}
# this file may define `proc foo`
source "my-foo.tcl" 
foo #; this now should resolve within "my-foo.tcl"
```

While we're still supporting only a single file, maybe we can just ignore this case. An alternative might be to just invalidate all definitions if we hit a `source` command, but that seems perhaps a bit too conservative. 

**Namespaces**

I think we may want to handle namespaces to some extent before making this functionality user visible. 

That will probably require tracking the namespace of each proc within the symbol table. There will also need to be code to parse fully-qualified names in `proc` definitions and references.

I think `namespace eval` is the most common method for defining/using namespaces, so we'll want to handle this. Perhaps the parser can tag `Script`nodes with a namespace (and maybe `CommandRef` nodes too?) and this info can be used by the symbol table/LSP server.

File: src/tclint/symbol_table.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    def __init__(self):
+        self.proc_def: DefaultDict[str, list[Node]] = defaultdict(list)
+
+    def add_proc_definition(self, command: Command) -> None:
+        """Add definition of procedure"""
+        # command holds the "proc" keyword, so the proc name is 1st argument
+        proc_name_node = command.args[0]
+        proc_name = proc_name_node.contents
+        assert proc_name is not None
```
[the remark this answers is not in the document]
Comment (patrislav1) [reply src/tclint/symbol_table.py]: resolved in https://github.com/nmoroze/tclint/pull/102/commits/7bde2bbd38b35e3bcbabf12ebbd77fd370cde7ab

File: src/tclint/symbol_table.py (hunk 2 above)
Comment (patrislav1) [reply src/tclint/symbol_table.py]: Agree that it's good to make the syntax tree smarter. If you like to do it pls go ahead.

Will this new node include both `proc` keyword and function name? (i.e. `pos` and `end_pos` will span the area of both words).

File: src/tclint/symbol_table.py
Comment (patrislav1) [reply src/tclint/symbol_table.py:1]: Resolved in https://github.com/nmoroze/tclint/pull/102/commits/7bde2bbd38b35e3bcbabf12ebbd77fd370cde7ab

File: src/tclint/syntax_tree.py [hunk 4]
```diff
@@ -232,6 +234,31 @@ def _recurse(self, visitor):
         for child in self.children:
             child.accept(visitor, recurse=True)
 
+    def _pos_match(self, line: int, col: int) -> bool:
+        """Return True if pos is within this node's block"""
+        if self.pos is None:
+            return False
+        if self.end_pos is None:
+            return line == self.pos[0] and col == self.pos[1]
+        # FIXME: check for off-by-one with end_pos
```
[the remark this answers is not in the document]
Comment (patrislav1) [reply src/tclint/syntax_tree.py:233]: Found and fixed/added test for off-by-one in https://github.com/nmoroze/tclint/pull/102/commits/5d9b39ead11587d257e0a7a5cbfbd222293dcba1

File: src/tclint/symbol_table.py (hunk 2 above)
Comment (nmoroze) [reply src/tclint/symbol_table.py]: > Will this new node include both proc keyword and function name? (i.e. pos and end_pos will span the area of both words).

I wasn't planning on it, looking at the Python LSP for example it doesn't seem like the function keyword actually triggers a go-to def or reference for the associated function. Would you expect different behavior?

Screenshot 2025-08-04 at 8 51 51 PM
